### PR TITLE
Enable PlugAndPlay for MQTT broker

### PR DESCRIPTION
**Pull request by dylanbronson**
Currently, PlugAndPlay devices cannot connect through Edge's MQTT Broker. This is because the broker does not store the modelId of devices/modules for later use.

This PR enables PlugAndPlay through the broker by storing any given ModelId for the identity in the MetadataStore when that identity is authenticated by EdgeHub Core. The broker will ask EH Core for authentication, and the EH Core's AuthAgentProtocolHead will do the authentication and store the ModelId in the MetadataStore, if is authenticated.

The metadataStore is already checked for a ModelId when EH is connecting to the cloud on behalf of an identity, so no change is needed there.

This change does NOT enable PlugAndPlay for the Mqtt Bridge, so nested scenario + Mqtt broker will NOT work yet. More work will need to be done to send the ModelId (and all of the username add-ons) upstream through the broker

Additionally, in a PR immediately after this one, I will add E2E tests for this case

### CODE REVIEW

File: edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter.Test/AuthAgentListenerTest.cs [hunk 1]
```diff
@@ -345,6 +346,31 @@ public async Task DeniesBadContentLengthLongBody()
...
+                content.version = "2020-04-20";
+                content.username = "testhub/device/api-version=2018-06-30&model-id=dtmi:test:modelId;1";
+                // [SuppressMessage("Microsoft.Security", "CS002:SecretInNextLine", Justification="Synthetic password used in tests")]
+                content.password = "somepassword";
+
+                dynamic response = await PostAsync(content, URL);
+
+                Assert.Equal(200, (int)response.result);
+                var modelId = (await metadataStore.GetMetadata("device")).ModelId;
+                Assert.True(modelId.HasValue);
```
Comment (vadim-kovalyov) [inline edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter.Test/AuthAgentListenerTest.cs:370]: nit: do you want to compare model id with what you set above?

File: edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter.Test/AuthAgentListenerTest.cs [hunk 2]
```diff
@@ -345,6 +346,31 @@ public async Task DeniesBadContentLengthLongBody()
...
+            IEntityStore<string, string> store = storeProvider.GetEntityStore<string, string>("productInfo");
+            var metadataStore = new MetadataStore(store, "productInfo");
+            using (var sut = new AuthAgentProtocolHead(authenticator, metadataStore, usernameParser, credFactory, sysIdProvider, config))
+            {
+                await sut.StartAsync();
+
+                dynamic content = new ExpandoObject();
+                content.version = "2020-04-20";
+                content.username = "testhub/device/api-version=2018-06-30&model-id=dtmi:test:modelId;1";
+                // [SuppressMessage("Microsoft.Security", "CS002:SecretInNextLine", Justification="Synthetic password used in tests")]
```
Comment (vadim-kovalyov) [inline edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter.Test/AuthAgentListenerTest.cs:363]: nit: is this still needed?

**Review by vadim-kovalyov — APPROVED** [review]

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/authentication/AuthAgentController.cs [hunk 3]
```diff
@@ -57,7 +59,8 @@ await this.GetIdsFromUsername(request)
                                     async creds => (await this.AuthenticateAsync(creds), Option.Some(creds)),
                                     () => Task.FromResult((false, Option.None<IClientCredentials>())));
 
-                return this.Json(GetAuthResult(isAuthenticated, credentials));
+                var authResult = await GetAuthResult(isAuthenticated, credentials, this.metadataStore);
```
Comment (vipeller) [inline edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/authentication/AuthAgentController.cs:62]: Would not be better to create a new StoreMetadata(credentials, this.metadataStore) method, and call that separately (if isAuthenticated == true), so then no need to change the GetAuthResult()? I feel it should not be the concern of GetAuthResult() to do other things than create the object that represents the result.

Comment (dylanbronson) [reply edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/authentication/AuthAgentController.cs:62]: Good call. I'll make the change.

File: edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter.Test/AuthAgentListenerTest.cs (hunk 1 above)
Comment (dylanbronson) [reply edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter.Test/AuthAgentListenerTest.cs:371]: oops forgot to do that

File: edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter.Test/AuthAgentListenerTest.cs (hunk 2 above)
Comment (dylanbronson) [reply edge-hub/core/test/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter.Test/AuthAgentListenerTest.cs:364]: Looks like it is based on the fact that it was committed like this for CredScan

**Review by vipeller — APPROVED** [review]

**Review by vadim-kovalyov — APPROVED** [review]

**Pull request merged**